compiler/rustc_codegen_llvm/src/mono_item.rs

```diff
@@ -62,8 +63,15 @@ impl<'tcx> PreDefineMethods<'tcx> for CodegenCx<'_, 'tcx> {
 
         let fn_abi = self.fn_abi_of_instance(instance, ty::List::empty());
         let lldecl = self.declare_fn(symbol_name, fn_abi, Some(instance));
-        unsafe { llvm::LLVMRustSetLinkage(lldecl, base::linkage_to_llvm(linkage)) };
         let attrs = self.tcx.codegen_fn_attrs(instance.def_id());
+        let llvm_linkage =
+            if attrs.flags.contains(CodegenFnAttrFlags::NAKED) && linkage == Linkage::Internal {
+                // this is effectively an extern fn, and must have external linkage
+                llvm::Linkage::ExternalLinkage
+            } else {
+                base::linkage_to_llvm(linkage)
+            };
+        unsafe { llvm::LLVMRustSetLinkage(lldecl, llvm_linkage) };
```

Comment:
this is very ugly, but required to link the extern fn with the global asm block. In practice, the value for `linkage` is used for both actual linkage as used here, and some form of visibility for the symbol. Setting the linkage to external earlier (e.g. in codegen_ssa when attributes are processed) is meaningfully different from doing so here (in that if it's set earlier, naked functions symbols are in the dynamic symbol table when they should not be). 

I suspect there is some better refactor/abstraction/idea here, but am not familiar enough with any of this code to see it.